introduce `SubPcGroup`

ThomasBreuer opened the pull request:
This is a first attempt to implement what was sketched in #2672.
I am not sure that we really want this:
- Admitting different types for a group and a subgroup was apparently not allowed up to now, several signatures of functions have to be "relaxed" in order to support different types.
- Functions returning arrays of subgroups (typically series such as `derived_series`) will now create the full group (if it occurs) with another type; one has to be aware of different types for the same mathematical object.

If we want to proceed like this then the next step is as follows.
- Fix functions like `intersection` which expect arguments of the same type up to now.
- Adjust the doctests.
- Adjust/extend the documentation.
- Add tests.
- Do the same for `FpGroup`/`SubFpGroup`.

ThomasBreuer committed:
introduce `SubPcGroup`

lgoettgens commented on src/Groups/GAPGroups.jl:
```diff
@@ -1065,7 +1066,7 @@ Return `C, f`, where `C` is the normal core of `H` in `G`,
 and `f` is the embedding morphism of `C` into `G`.
 """
-core(G::T, H::T) where T<:GAPGroup = _as_subgroup(G, GAPWrap.Core(G.X, H.X))
+core(G::T1, H::T2) where {T1<:GAPGroup, T2<:GAPGroup} = _as_subgroup(G, GAPWrap.Core(G.X, H.X))
```
```suggestion
core(G::GAPGroup, H::GAPGroup) = _as_subgroup(G, GAPWrap.Core(G.X, H.X))
```
Your change makes it actually a lot easier

fingolfin replied:
It makes the signatures easier (with Lars' extra change). Of course we loose a bit of "validation", i.e. previously Julia prevented some situations from occurring (e.g. you could not mix a permutation group with a matrix group), and now we have to catch that ourselves. But that's not so bad in the end, we already have to catch some of these "bad" situations, e.g. when both groups are matrix groups but defined over different rings (perhaps in different characteristic)

fingolfin commented:
> This is a first attempt to implement what was sketched in #2672. I am not sure that we really want this:
> 
> * Admitting different types for a group and a subgroup was apparently not allowed up to now, several signatures of functions have to be "relaxed" in order to support different types.

That seems fine to me.

> * Functions returning arrays of subgroups (typically series such as `derived_series`) will now create the full group (if it occurs) with another type; one has to be aware of different types for the same mathematical object.

I also think this is fine, and in fact just reflects the reality in GAP, which tries to paper of it but doesn't really succeed. So I actually think it is *better* that we make this explicit. This way at least the user can immediately tell if a given group is the "full" group or not, and we have a chance of explaining the situation in the docstring for `SubPcGroup`.

> 
> If we want to proceed like this then the next step is as follows.
> 
> * Fix functions like `intersection` which expect arguments of the same type up to now.
> * Adjust the doctests.
> * Adjust/extend the documentation.
> * Add tests.
> * Do the same for `FpGroup`/`SubFpGroup`.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -321,10 +321,11 @@ function Base.show(io::IO, G::PermGroup)
   @show_name(io, G)
   @show_special(io, G)
-  print(io, "Pc group")
+  T = typeof(G) == PcGroup ? "Pc group" : "Sub pc group"
```
Maybe like this?
```suggestion
  T = typeof(G) == PcGroup ? "Pc group" : "Subgroup of pc group"
```

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -110,11 +117,12 @@ function cyclic_group(::Type{T}, n::Union{IntegerUnion,PosInf}) where T <: GAPGr
 
-function cyclic_group(::Type{PcGroup}, n::Union{IntegerUnion,PosInf})
+# special handling for pc groups: distinguish on the GAP side
+function cyclic_group(::Type{T}, n::Union{IntegerUnion,PosInf}) where T <: Union{PcGroup, SubPcGroup}
```
This creates a new `PcGroup`, which would be the "whole" group, so should we really allow `SubPcGroup` here?

fingolfin replied:
@ThomasBreuer did you see my question from January there? I don't terribly mind having this (it may make some generic code easier to write), but I wonder: does `cyclic_group(SubPcGroup, 5)` really work correctly, and return something of type `SubPcGroup`? We don't have a test for that, do we?

ThomasBreuer replied:
Yes, `cyclic_group(SubPcGroup, 5)` works.
(I had intended to add a test for this, but somehow this got lost.)

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -517,6 +524,29 @@ function isomorphism(::Type{T}, G::GAPGroup) where T <: Union{FPGroup, PcGroup,
    end::GAPGroupHomomorphism{typeof(G), T}
 end
 
+# If `G` is not a full pc group then switch to a full pc group in Oscar.
```
This, by the way, is something that I needed a lot in the past, and was always annoyingly complicated in GAP. So this is a nice bonus I did not even consider (at least I don't recall considering it :-) )

ThomasBreuer replied:
I think GAP is just missing something like `IsomorphismFullPcGroup`.

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -517,6 +524,29 @@ function isomorphism(::Type{T}, G::GAPGroup) where T <: Union{FPGroup, PcGroup,
+     # The codomain of `f` can be a *subgroup* of a full pc group.
+     # In this situation, we have to switch to a full pc group.
+     C = GAP.Globals.Range(f)::GapObj
+     if GAPWrap.GeneratorsOfGroup(C) != GAP.Globals.FamilyPcgs(C)
```
What about `pcp` groups, they don't use `FamilyPcgs` and probably needs separate code?

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -741,12 +771,14 @@ end
     PcGroup(G::T) where T <: Union{GAPGroup, GrpAbFinGen}
     pc_group(G::T) where T <: Union{GAPGroup, GrpAbFinGen}
+    SubPcGroup(G::T) where T <: Union{GAPGroup, GrpAbFinGen}
+    sub_pc_group(G::T) where T <: Union{GAPGroup, GrpAbFinGen}
```
Do we really want/need this? Maybe `SubPcGroup` on a `PcGroup` could be useful, in case one needs the "full group as a subgroup", but other than that?

fingolfin commented on src/Groups/types.jl:
````diff
@@ -197,6 +229,53 @@ as shown in Section [Elements of groups](@ref elements_of_groups).
+
+```
+"""
+const SubPcGroupElem = BasicGAPGroupElem{SubPcGroup}
````
To me the most annoying part of this PR actually seems to be `SubPcGroupElem`; we probably would want to allow multiplying a `SubPcGroupElem` by a `PcGroupElem` (and vice-versa), and possibly in various other contexts one may have to / wish to support both?

Perhaps a suitable `promote_rule` can help a bit?

ThomasBreuer committed:
address comments

fingolfin commented on src/Groups/sub.jl:
````diff
@@ -850,7 +858,7 @@ julia> typeof(quo(PermGroup, G, N)[1])
 ```
 """
-function quo(G::T, N::T) where T <: GAPGroup
+function quo(G::T1, N::T2) where {T1 <: GAPGroup, T2 <: GAPGroup}
````
We could add a helper
```julia
_is_subgroup_type(::Type{T1}, ::Type{T2}) where {T1, T2} = false
_is_subgroup_type(::Type{T}, ::Type{T}) where {T <: GAPGroup} = true
_is_subgroup_type(::Type{PcGroup}, ::Type{SubPcGroup}) = true
```
and then add checks like
```julia
  @req _is_subgroup_type(T1, T2) "argument types do not fit"
```

ThomasBreuer replied:
I thought about introducing something like `_compatibility_check(G1::T1, G2::T2)` which throws an exception if the arguments do not fit together, and otherwise has no effect. Note that checking the types alone is not sufficient, for example in the case of two f.p. subgroups or two pc subgroups which belong to different full groups. If we do not check this compatibility before delegating to GAP then we get just a cryptic error message from GAP.

fingolfin commented:
Just to document it somewhere in writing: We discussed this some time ago and the verdict was that yes, we want to proceed in this direction; while there are downsides, many of them actually already exists, but just are hidden; and there are also upsides. So...

fingolfin commented:
@ThomasBreuer I hope this is still on your radar? :-)

ThomasBreuer committed:
next step towards `SubPcGroup`

- add `_check_compatible`
- remove requirements that several groups as arguments
  must have the same type; instead call `_check_compatible` inside

ThomasBreuer committed:
lots of changes, now at least the tests pass again

- add new GAP property `GroupGeneratorsDefinePresentation`,
  in order to check/cache whether a GAP group is the full group
  given by the underlying presentation, *and* whether the stored
  group generators correspond to the generators of this presentation

- export `SubPcGroup`, `SubPcGroupElem`

- relax type relations in various signatures,
  since now subgroups of a group can have a type
  different from that of the full group

- change the parametrization of various types
  because now several types of groups may be involved:
  `GroupCoset` involves a group, a subgroup, and a group element.
  `GroupDoubleCoset` involves a group, two subgroups, and a group element.
  `SubgroupTransversal` involves a group and a subgroup

- add a dictionary `_sub_types` that maps a group type to the type
  that is used for its (not necessariy proper) subgroups,
  adjust `_oscar_group` to create a group of the subgroup type
  (this is what `_oscar_group` is thought for),
  adjust `_get_type`,

- support `on_gens = true` also for `isomorphism(PcGroup, G)`
  where possible

- change some (types of) return values:
  - the default return type of `_isomorphic_gap_group(A::FinGenAbGroup)`
    from `PcGroup` to `SubPcGroup`,
  - the return value of `isomorphism_to_GAP_group(G::FinGenAbGroup)`
    to use a `SubPcGroup` not a `PcGroup`,
  - throw an error if `abelian_group(PcGroup, G)` is called for a group
    `G` whose generators do not fit to a supported pc presentation,
  - and throw an error if `isomorphism(PcGroup, A)` is called for
    an abelian group such that the generators of `A` do not correspond
    to the generators of the underlying presentation.

  Yes, this is not a good idea.

  The problem is that currently we cannot guarantee that the `gens`
  value of a newly constructed `PcGroup` isomorphic to a given abelian group
  corresponds to the generators in the defining presentation.
  (GAP's `IsPcGroup` groups support only prime relative orders,
  and we cannot use GAP's `IsPcpGroup` throughout because of bugs that
  show up in Oscar's CI tests.)

  (Question:
  Should the constructor `PcGroup(G::GapObj)`,
  where `G` is a non-full pc group in GAP, be allowed to replace `G` be a
  newly created independent group, or should it throw an error,
  and `pc_group` should do the switch to a new group before calling `PcGroup`?)

- disable (temporarily) some `@inferred` calls in tests,
  due to situations where types of results depend on `_sub_types`

- call `abelian_group(SubPcGroup, ...)` instead of `abelian_group(PcGroup, ...)`
  in several tests, where the latter is currently not supported
  (hopefully temporarily);
  analogously for `isomorphism(PcGroup, ...)` vs. `isomorphism(SubPcGroup, ...)`

- disable (temporarily) two equality tests concerning mappings from/to
  `PcGroup`s or `SubPcGroup`s:
  Would it be o.k. to just regard the groups (the maps) as equal,
  independent of their types?

ThomasBreuer committed:
`@inferred` tests work again

due to changed `sub_type` handling

ThomasBreuer committed:
adjust docstrings to signature changes

ThomasBreuer committed:
adjust serialization to type `SubPcGroup`

ThomasBreuer committed:
more adjustments

- adjust doctest output to `SubPcGroup` objects
- changed signature of `complements`, `complement_classes`, added tests
- document `SubPcGroupElem`

ThomasBreuer committed:
fix relation between `PcGroup` and `pc_group`

`PcGroup` takes a GAP group and wraps it;
it is not allowed to replace the group by a different GAP group,
thus it must throw an exception if the GAP group is not suitable.
(This is crucial for example in the creation of group homomorphisms,
where the GAP objects stored in domain and codomain must fit to the
`Source` and `Range` objects of the underlying GAP mapping.)

`pc_group` may return an Oscar group that stores a different
GAP group than the given one.

ThomasBreuer committed:
fix `isomorphism(PcGroup, A::FinGenAbGroup)`

The function must be able to create a map to a `PcGroup`,
also in the case that `abelian_group(PcGroup, exponents)` does not work,
due to the requirements that the `gens` value of the result
corresponds to the given element orders *and* is the defining
polycyclic sequence of the result group.

ThomasBreuer committed:
use an embedding

This fixes a test.
I am not sure whether we want a more strict use of embeddings,
or to allow for a more sloppy use of group elements.
(In the latter case, we will run into problems in examples
involving `FinGenAbGroup`s.)

ThomasBreuer committed:
adjust two doctests

ThomasBreuer committed:
more fixes

- fix a signature in a `.md` file
- remove a type assertion that is no longer valid
  (a subgroup can have a different type)

ThomasBreuer commented:
The failure of one of the book examples shows a conceptual problem. (Likely this example can be turned back to the old behaviour by an improvement on the GAP side, but this does not solve the general problem.)

The function `quo` for two groups states that the type of the return value depends on the situation: Anyhow a new group must be created, and GAP decides about a suitable representation; if the result is known to be a finite solvable group then a pc group is returned, otherwise a permutation group or a fp group are possible.

The GAP function `NaturalHomomorphismByNormalSubgroup` is allowed to return a mapping that is not (known to be) surjective. If this happens and if the `Range` of the mapping is a pc group then the quotient on the Oscar side can be only a `SubPcGroup`. (At some extra cost, we can check the surjectivity, and create a `PcGroup` in Oscar if yes.)

It looks as if we have to do something more specific on the GAP side if we want the results of `quo` to be either `PcGroup` or `FPGroup` or `PermGroup`.

(@fieker @fingolfin ?)

ThomasBreuer committed:
fix cross-references

ThomasBreuer committed:
get rid of `_get_type`

ThomasBreuer committed:
change `quo` to prefer a `PcGroup` result to a `SubPcGroup`

I am not sure whether this is a good idea.

On the one hand, users would not understand that a newly created
quotient is represented as a `SubPcGroup`.
(The starting point was changed output in an example for the Oscar book.)

On the other hand, we have to work around a GAP feature that offers
natural homomorphisms instead of natural epimorphisms (with the
possibility to create maps that need not be surjective).

Since we have to compute the image of the GAP mapping anyhow,
the current change does not add expensive computations.

ThomasBreuer commented:
The failure in `test (1.10, book, ubuntu-latest)` is due to an intended change of the output shown: The group in question has now a different type.

benlorenz commented:
> The failure in `test (1.10, book, ubuntu-latest)` is due to an intended change of the output shown: The group in question has now a different type.

Then you can adjust the output in (`test/book/cornerstones/groups/actions.jlcon`) to the new printing. Only the input must stay identical for 1.1 (i.e. current master).

ThomasBreuer committed:
adjust output in an Oscar Book example

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -568,6 +576,57 @@ function isomorphism(::Type{FPGroup}, G::GAPGroup; on_gens::Bool=false)
+         f = GAP.Globals.GroupHomomorphismByImages(G, CC, Cpcgs, CCpcgs)::GapObj
+         return GAPGroupHomomorphism(G, T(CC), f)
+       else
+         f = GAP.Globals.IsomorphismPcpGroup(G.X)::GapObj
```
```suggestion
         f = GAP.Globals.IsomorphismPcpGroup(GapObj(G))::GapObj
```
possibly also similar (i.e. `FOO.X` to `GapObj(FOO)`) elsewhere?

ThomasBreuer replied:
O.k., I change all those `.X` introduced in this pull request; if I change also others then there will likely be conflicts.

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -568,6 +576,57 @@ function isomorphism(::Type{FPGroup}, G::GAPGroup; on_gens::Bool=false)
+       # and a group in `IsPcpGroup` otherwise.
+       if is_finite(G)
+         fam = GAP.Globals.ElementsFamily(GAP.Globals.FamilyObj(G.X))::GapObj
+         Ggens = GapObj([x.X for x in gens(G)])::GapObj
```
Huh, does this line work as intended? There is no test case covering this code according to Codecov, and I have some doubts: won't this create a GAP list of OSCAR elements?

Perhaps like this:
```suggestion
         Ggens = GapObj(gens(G); recursive=true)
```

ThomasBreuer replied:
`gens(G)` consists of OSCAR elements, the array of the `x.X` consists of GAP elements.
I am going to add tests.
(And yes, the code you propose is better.)

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -622,18 +681,32 @@ function isomorphism(::Type{T}, A::FinGenAbGroup) where T <: GAPGroup
+     # `T == PcGroup`, hence we cannot call `abelian_group(T, exponents)`.)
+     if T == PcGroup
+       if 0 in exponents
+         GapG = GAP.Globals.AbelianPcpGroup(length(exponents), GAP.GapObj(exponents, recursive=true))
```
I guess we can drop the `GAP.` from `GAP.GapObj` here and elsewhere? Then again we still use it in a bunch of places, so we could change that in a later follow-up PR.
```suggestion
         GapG = GAP.Globals.AbelianPcpGroup(length(exponents), GapObj(exponents, recursive=true))
```

ThomasBreuer replied:
Yes, we can write just `GapObj` because meanwhile it is an exported variable.

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -811,12 +885,13 @@ end
-@attributes MultTableGroup
+# # Remove once Hecke lower bound is >= 0.14.1
+# @attributes MultTableGroup
 
```
```suggestion
```

fingolfin commented on src/Groups/matrices/MatGrp.jl:
```diff
@@ -435,6 +435,7 @@ function Base.:^(x::MatrixGroupElem, y::MatrixGroupElem)
 end
 
 comm(x::MatrixGroupElem, y::MatrixGroupElem) = inv(x)*conj(x,y)
+#T why needed? GAPGroupElem has x^-1*x^y
```
If we don't need it, remove it? (Though perhaps in a separate PR?)

Though then again, now I wonder if the general one shouldn't be re-implemented by calling `GAP.Globals.Comm`? And then it might make sense to keep the above method here... Or even change it to, say, `inv(y*x)*(x*y)` (one less inversion, which is arguably the most expensive bit here?)

ThomasBreuer replied:
Let us leave this to another pull request.

fingolfin commented on src/Groups/sub.jl:
```diff
@@ -175,13 +188,13 @@ julia> normal_subgroups(symmetric_group(5))
+6-element Vector{SubPcGroup}:
+ Subgroup of pc group of order 8
+ Subgroup of pc group of order 4
+ Subgroup of pc group of order 4
```
Hmmm... this is ambiguous, isn't it? The `4` refers to the order of the subgroup, but a reader might reasonably interpret this as saying the parent group has order 4... :-/.

(How) can we make this clearer? Perhaps We should call it a `Sub-pc group` (and explain that somewhere)? 

Or use a different phrasing: `Subgroup of pc group, has order 4` -- alas that would be different from how this kind of info is printed for almost anything else in OSCAR, wouldn't it?

Perhaps @thofma or @fieker have some brilliant insight on this.

fingolfin replied:
To be clear: we can address this in a follow-up PR (but then let's make an issue about this after the PR is merged so so we don't forget)

ThomasBreuer replied:
I have opened issue  #3732.

fingolfin commented on src/Groups/sub.jl:
```diff
@@ -792,20 +810,26 @@ julia> typeof(quo(PermGroup, G, N)[1])
+#T       where possible, take the *full* pc group if the range is a pc group
+#T       and if the GAP mapping is surjective.
+  cod = GAP.Globals.Range(mp)::GapObj
+  if !(GAP.Globals.IsPcGroup(cod) && GAP.Globals.IsSurjective(mp))
```
```suggestion
  if !(GAPWrap.IsPcGroup(cod) && GAPWrap.IsSurjective(mp))
```

fingolfin commented on src/Groups/sub.jl:
```diff
@@ -792,20 +810,26 @@ julia> typeof(quo(PermGroup, G, N)[1])
+#T HACK: In order to avoid `SubPcGroup`s as return values of `quo`
+#T       where possible, take the *full* pc group if the range is a pc group
+#T       and if the GAP mapping is surjective.
+  cod = GAP.Globals.Range(mp)::GapObj
```
```suggestion
  cod = GAPWrap.Range(mp)
```

fingolfin commented on src/Groups/sub.jl:
```diff
@@ -792,20 +810,26 @@ julia> typeof(quo(PermGroup, G, N)[1])
+#T       and if the GAP mapping is surjective.
+  cod = GAP.Globals.Range(mp)::GapObj
+  if !(GAP.Globals.IsPcGroup(cod) && GAP.Globals.IsSurjective(mp))
+    cod = GAP.Globals.ImagesSource(mp)::GapObj
```
Maybe "GAP wrap" `ImagesSource`, too?

fingolfin commented on src/Groups/types.jl:
```diff
@@ -275,8 +357,19 @@ end
+  @assert GAP.Globals.FamilyObj(S.X) === GAP.Globals.FamilyObj(G.X)
+  return S
+end
+#T better rename to _oscar_subgroup?
```
I agree: `_oscar_subgroup` would be better

ThomasBreuer replied:
O.k., I will rename the two-argument version.

fingolfin commented on src/Groups/types.jl:
```diff
@@ -302,10 +395,12 @@ function _oscar_group(obj::GapObj, G::MatrixGroup)
 # "Coerce" an Oscar group `G` to one that is compatible with
 # the given Oscar group `S`.
+#T what does compatible mean?
 compatible_group(G::T, S::T) where T <: GAPGroup = _oscar_group(GapObj(G), S)
```
This is not used anywhere, is it? Can we just remove it?

ThomasBreuer replied:
Yes.

fingolfin commented on test/Groups/homomorphisms.jl:
```diff
@@ -580,8 +582,9 @@ function test_kernel(G,H,f)
 
-   @test preimage(f,H)==(G,id_hom(G))
-   @test preimage(f,sub(H,[one(H)])[1])==(K,i)
+#TODO: activate these tests as soon as they pass again
```
What is the issue here? Can we perhaps at least record that in the comment?
I.e. do these run into errors? Or return wrong results? Or something else? If there is a relevant issue, link to it here?

ThomasBreuer replied:
I will add a comment.

fingolfin commented on test/Groups/libraries.jl:
```diff
@@ -139,7 +139,8 @@ end
 @testset "Small groups" begin
    L = all_small_groups(8)
-   LG = [abelian_group(PcGroup,[2,4]), abelian_group(PcGroup,[2,2,2]), cyclic_group(8), quaternion_group(8), dihedral_group(8)]
+#  LG = [abelian_group(PcGroup,[2,4]), abelian_group(PcGroup,[2,2,2]), cyclic_group(8), quaternion_group(8), dihedral_group(8)]
```
Is there a point in keep the "old" copies of the test around, here and below? Otherwise just delete them?

ThomasBreuer replied:
I am still confident that eventually we can support `abelian_group(PcGroup, [2, 4])`.
I will add a comment.

fingolfin commented on test/Groups/quotients.jl:
```diff
@@ -52,14 +53,16 @@ end
    #    return a group of this type if possible
-   for T in [ PermGroup, PcGroup ]
+#  for T in [ PermGroup, PcGroup ]
+   for T in [ PermGroup, SubPcGroup ]
```
Why not
```suggestion
   for T in [ PermGroup, PcGroup, SubPcGroup ]
```

ThomasBreuer replied:
The point is that currently `abelian_group(PcGroup, [2, 3, 4])` is not supported.
`abelian_group` promises a result whose generators correspond to the given orders,
and the type `PcGroup` implies that the generators are a polycyclic generating sequence w.r.t. the underlying presentation.
With GAP's `IsPcGroup`, we cannot get this; and `IsPcpGroup` cannot be used because of some bugs.

(I will add a comment.)

fingolfin reviewed (commented):
Some remarks. Most are optional (better to get this merged sooner and then polish afterwards), but there are one or two were I wonder if the code is correct as it is...?

ThomasBreuer committed:
address reviewer's remarks

- add comments
- add tests
- improve error messages

fingolfin commented:
Unfortunately this has conflicts again now :-(

It would be great if we could merge this tomorrow (before or after triage, whenever you have time to resolve it) so we can move on :-)

ThomasBreuer commented:
> Unfortunately this has conflicts again now :-(

I find it rather surprising that there were not more conflicts.

fingolfin commented:
@ThomasBreuer there is one unresolved review comment of mine from quite some ago where I ask about `sub_pc_group` -- can you please reply to that?
